core/src/header_chain_prover.rs

```diff
@@ -0,0 +1,718 @@
+            Network::Bitcoin => MAINNET_ELF.as_ref(),
+            Network::Testnet => TESTNET4_ELF.as_ref(),
+            Network::Testnet4 => TESTNET4_ELF.as_ref(),
+            Network::Signet => SIGNET_ELF.as_ref(),
+            Network::Regtest => REGTEST_ELF.as_ref(),
+            _ => Err(BridgeError::UnsupportedNetwork.into_eyre())?,
+        };
+
+        let receipt = prover.prove(env, elf).map_err(|e| eyre::eyre!(e))?.receipt;
+        tracing::debug!(
+            "Proof receipt for header chain circuit input {:?}: {:?}",
+            input,
+            receipt
+        );
+
+        Ok(receipt)
+    }
+
+    #[tracing::instrument(skip_all)]
+    pub async fn check_for_new_unproven_blocks(
+        &self,
+    ) -> Result<Option<(BlockHash, Header, u32, Receipt)>, BridgeError> {
+        let non_proved_block = self.db.get_non_proven_block(None).await;
+
+        match non_proved_block {
+            Ok(non_proved_block) => Ok(Some((
+                non_proved_block.0,
+                non_proved_block.1,
+                non_proved_block
+                    .2
+                    .try_into()
+                    .wrap_err("Can't convert i32 to u32")?,
+                non_proved_block.3,
+            ))),
+            Err(BridgeError::DatabaseError(sqlx::Error::RowNotFound)) => Ok(None),
+            Err(e) => Err(e),
+        }
+    }
+
+    #[tracing::instrument(skip_all)]
+    pub async fn prove_block(
+        &self,
+        current_block_hash: BlockHash,
+        current_block_header: Header,
+        current_block_height: u32,
+        previous_proof: Receipt,
+    ) -> Result<Receipt, BridgeError> {
+        tracing::info!(
+            "Prover starts proving for block with hash {} and with height {}",
+            current_block_hash,
+            current_block_height
+        );
+
+        let header: CircuitBlockHeader = current_block_header.into();
+        let receipt = self.prove_block_headers(Some(previous_proof), vec![header.clone()])?;
+
+        self.db
+            .set_block_proof(None, current_block_hash, receipt.clone())
+            .await?;
+
+        Ok(receipt)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct HeaderChainProverClient {
+    db: Database,
+}
+
+impl HeaderChainProverClient {
+    pub async fn new(db: Database) -> Result<Self, HeaderChainProverError> {
+        Ok(HeaderChainProverClient { db })
+    }
+
+    /// Get the proof of a block.
+    ///
+    /// # Parameters
+    ///
+    /// - `hash`: Target block hash
+    ///
+    /// # Returns
+    ///
+    /// - [`Receipt`]: Specified block's proof receipt
+    pub async fn get_header_chain_proof(&self, hash: BlockHash) -> Result<Receipt, BridgeError> {
+        match self.db.get_block_proof_by_hash(None, hash).await? {
+            Some(r) => Ok(r),
+            None => Err(HeaderChainProverError::NoHeaderChainProof(hash).into()),
+        }
+    }
+
+    /// Sets a new block to database, later to be proven.
```

Comment:
"later to be proven" as in "the blockgazer will prove the saved block" right? also "save_new_block" is better for this

Reply:
yeah sorry, I meant to say "the header chain prover will prove the saved block in its task"